File: cpp/leetcode/editor/cn/[360]有序转化数组.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>
#include <utility>
#include <memory>
#include <algorithm>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)

class IterHelper {
public:
    IterHelper(vector<int> &vec, bool reverse) : _reverse(reverse) {
        if (reverse)
            _right = vec.rbegin();
        else
            _left = vec.begin();
    }

    void operator++(int n) {
        if (_reverse)
            _right++;
        else
            _left++;
    }

    int &operator*() {
        if (_reverse)
            return *_right;
        else
            return *_left;
    }

    bool _reverse{};
    vector<int>::iterator _left;
    vector<int>::reverse_iterator _right;
};

class Solution {
public:
    vector<int> sortTransformedArray(vector<int> &nums, int a, int b, int c) {
        vector<int> res(nums.size());

        if (a == 0) {
            if (b >= 0) {
                transform(nums.begin(), nums.end(),
                          res.begin(),
                          [&b, &c](int x) { return b * x + c; });
                return res;
            } else {
                transform(nums.begin(), nums.end(), res.rbegin(), [&b, &c](int x) { return b * x + c; });
                return res;
            }
        } else {
            float mid = -(float) b / (2.0f * (float) a);
            auto left = nums.begin();
            auto right = nums.end() - 1;
            auto p = IterHelper(res, a > 0);

            while (left <= right) {
                if (abs((float) *left - mid) > abs((float) *right - mid)) {
                    *p = a * (*left) * (*left) + b * (*left) + c;
                    left++;
                } else {
                    *p = a * (*right) * (*right) + b * (*right) + c;
                    right--;
                }
                p++;
            }
            return res;
        }
    }

};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
    Solution s;
    vector<int> v = {-4, -2, 2, 4};
    s.sortTransformedArray(v, 1, 3, 5);
}

File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[360]有序转化数组.scala
package leetcode.editor.cn
object SortTransformedArray {
//leetcode submit region begin(Prohibit modification and deletion)
  object Solution {
    def sortTransformedArray(nums: Array[Int], a: Int, b: Int, c: Int): Array[Int] =
      val res = Array.ofDim[Int](nums.length)
      val vertHyperbola = if (a > 0) true else false
      var p = if (vertHyperbola) nums.length - 1 else 0
      var p1 = 0
      var p2 = nums.length - 1

      while p1 <= p2 do
        val f1 = f(nums(p1), a, b, c)
        val f2 = f(nums(p2), a, b, c)
        if vertHyperbola
        then
          if f1 > f2
          then
            res(p) = f1
            p1 += 1
          else
            res(p) = f2
            p2 -= 1
          p -= 1
        else
          if f1 > f2
          then
            res(p) = f2
            p2 -= 1
          else
            res(p) = f1
            p1 += 1
          p += 1
      res

    private def f(x: Int, a: Int, b: Int, c: Int): Int =
      a * x * x + b * x + c
  }
//leetcode submit region end(Prohibit modification and deletion)

}
